services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/KnownContentLengthAsyncRequestBodySubscriber.java

```diff
@@ -0,0 +1,220 @@
+                                                              Pair.of(uploadRequest, asyncRequestBody))
+                             .whenComplete((r, t) -> {
+                                 if (t != null) {
+                                     if (shouldFailRequest()) {
+                                         multipartUploadHelper.failRequestsElegantly(futures, t, uploadId, returnFuture,
+                                                                                     putObjectRequest);
+                                     }
+                                 } else {
+                                     completeMultipartUploadIfFinished(asyncRequestBodyInFlight.decrementAndGet());
+                                 }
+                             });
+        subscription.request(1);
+    }
+
+    private boolean shouldFailRequest() {
+        return failureActionInitiated.compareAndSet(false, true) && !isPaused;
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        log.debug(() -> "Received onError ", t);
+        if (failureActionInitiated.compareAndSet(false, true)) {
+            multipartUploadHelper.failRequestsElegantly(futures, t, uploadId, returnFuture, putObjectRequest);
+        }
+    }
+
+    @Override
+    public void onComplete() {
+        log.debug(() -> "Received onComplete()");
+        isDone = true;
+        if (!isPaused) {
+            completeMultipartUploadIfFinished(asyncRequestBodyInFlight.get());
+        }
+    }
+
+    private void completeMultipartUploadIfFinished(int requestsInFlight) {
+        if (isDone && requestsInFlight == 0) {
+            CompletedPart[] parts;
+            if (existingParts.isEmpty()) {
+                parts = completedParts.values().toArray(new CompletedPart[0]);
+            } else if (!completedParts.isEmpty()) {
+                // List of CompletedParts needs to be in ascending order
+                parts = mergeCompletedParts();
+            } else {
+                parts = existingParts.values().toArray(new CompletedPart[0]);
+            }
+            completeMpuFuture = multipartUploadHelper.completeMultipartUpload(returnFuture, uploadId, parts,
+                                                                              putObjectRequest);
+        }
+    }
+
+    private CompletedPart[] mergeCompletedParts() {
+        CompletedPart[] merged = new CompletedPart[partCount];
+        int currPart = 1;
+        while (currPart < partCount + 1) {
+            CompletedPart completedPart = existingParts.containsKey(currPart) ? existingParts.get(currPart) :
+                                          completedParts.get(currPart);
+            merged[currPart - 1] = completedPart;
+            currPart++;
+        }
+        return merged;
+    }
+
+    @SdkTestInternalApi
+    public void setCompleteMpuFuture(CompletableFuture<CompleteMultipartUploadResponse> completeMpuFuture) {
```

Comment:
Same here, suggesting using dependency injection for testing instead (passing a mock `multipartUploadHelper` that returns a mocked completeMpuFuture)